File: leetcode/linked_list/1634_Add_Two_Polynomials_Represented_as_Linked_Lists/codes/ll.cpp
/**
Definition for polynomial singly-linked list.
 */
struct PolyNode {
    int coefficient, power;
    PolyNode *next;
    PolyNode(): coefficient(0), power(0), next(nullptr) {}
    PolyNode(int x, int y): coefficient(x), power(y), next(nullptr) {}
    PolyNode(int x, int y, PolyNode* next): coefficient(x), power(y), next(next) {}
};

class Solution {
public:
    PolyNode* addPoly(PolyNode* poly1, PolyNode* poly2) {
        PolyNode* dummy = new PolyNode();
        PolyNode* curr = dummy;
        while (poly1 && poly2) {
            if (poly1 -> power == poly2 -> power) {
                int coef = poly1 -> coefficient + poly2 -> coefficient;
                if (coef != 0) {
                    curr -> next = new PolyNode(coef, poly1 -> power);
                    curr = curr -> next;
                }
                poly1 = poly1 -> next;
                poly2 = poly2 -> next;
            } else if (poly1 -> power > poly2 -> power) {
                curr -> next = new PolyNode(poly1 -> coefficient, poly1 -> power);
                curr = curr -> next;
                poly1 = poly1 -> next;
            } else {
                curr -> next = new PolyNode(poly2 -> coefficient, poly2 -> power);
                curr = curr -> next;
                poly2 = poly2 -> next;
            }
        }
        while (poly1) {
            curr -> next = new PolyNode(poly1 -> coefficient, poly1 -> power);
            curr = curr -> next;
            poly1 = poly1 -> next;
        }
        while (poly2) {
            curr -> next = new PolyNode(poly2 -> coefficient, poly2 -> power);
            curr = curr -> next;
            poly2 = poly2 -> next;
        }
        return dummy -> next;
    }
};

File: leetcode/linked_list/1634_Add_Two_Polynomials_Represented_as_Linked_Lists/codes/ll.py
# Definition for polynomial singly-linked list
class PolyNode:
    def __init__(self, x=0, y=0, next=None):
        self.coefficient = x
        self.power = y
        self.next = next

class Solution:
    def addPoly(self, poly1: 'PolyNode', poly2: 'PolyNode') -> 'PolyNode':
        dummy = PolyNode()
        curr = dummy
        while poly1 and poly2:
            if poly1.power == poly2.power:
                coef = poly1.coefficient + poly2.coefficient
                if coef != 0:
                    curr.next = PolyNode(coef, poly1.power)
                    curr = curr.next
                poly1 = poly1.next
                poly2 = poly2.next
            elif poly1.power > poly2.power:
                curr.next = PolyNode(poly1.coefficient, poly1.power)
                curr = curr.next
                poly1 = poly1.next
            else:
                curr.next = PolyNode(poly2.coefficient, poly2.power)
                curr = curr.next
                poly2 = poly2.next
        while poly1:
            curr.next = PolyNode(poly1.coefficient, poly1.power)
            curr = curr.next
            poly1 = poly1.next
        while poly2:
            curr.next = PolyNode(poly2.coefficient, poly2.power)
            curr = curr.next
            poly2 = poly2.next
        return dummy.next
